### PR TITLE
Add performUpkeep simulation to keepers in separate eth_call

### PR DESCRIPTION
**Pull request by cmalec**
https://app.shortcut.com/chainlinklabs/story/19393/add-performupkeep-simulation-to-keepers-in-separate-eth-call

- adding 3 steps to pipeline for keepers
-- check perform does a simulated perform
-- decode data for above
-- conditional that checks if the perform returned a success
- new task type conditional added that currently just checks success value
- task eth call now has a from address added for perform check
- some variable name changes and comments for code readability

### CODE REVIEW

File: core/services/keeper/validate.go [hunk 1]
```diff
@@ -33,6 +33,19 @@ decode_check_upkeep_tx   [type=ethabidecode
 encode_perform_upkeep_tx [type=ethabiencode
                           abi="performUpkeep(uint256 id, bytes calldata performData)"
                           data="{\"id\": $(jobSpec.upkeepID),\"performData\":$(decode_check_upkeep_tx.performData)}"]
+check_perform_upkeep_tx  [type=ethcall
```
Comment (RyanRHall) [inline core/services/keeper/validate.go:36]: to avoid confusion, could we maybe call this step `simulate_perform_upkeep_tx`?

File: core/services/keeper/validate.go [hunk 2]
```diff
@@ -33,6 +33,19 @@ decode_check_upkeep_tx   [type=ethabidecode
...
+                          failEarly=true
+                          extractRevertReason=true
+                          evmChainID="$(jobSpec.evmChainID)"
+                          contract="$(jobSpec.contractAddress)"
+                          from="$(jobSpec.fromAddress)"
+                          gas="$(jobSpec.performUpkeepGasLimit)"
+                          data="$(encode_perform_upkeep_tx)"]
+decode_check_perform_tx  [type=ethabidecode
+                          abi="bool success"]
+conditional_tx           [type=conditional
```
Comment (RyanRHall) [inline core/services/keeper/validate.go:46]: `conditional_tx` --> the "tx" part doesn't make sense to me. Maybe `check_success` or something?

File: core/services/keeper/validate.go [hunk 3]
```diff
@@ -33,6 +33,19 @@ decode_check_upkeep_tx   [type=ethabidecode
 encode_perform_upkeep_tx [type=ethabiencode
                           abi="performUpkeep(uint256 id, bytes calldata performData)"
                           data="{\"id\": $(jobSpec.upkeepID),\"performData\":$(decode_check_upkeep_tx.performData)}"]
+check_perform_upkeep_tx  [type=ethcall
+                          failEarly=true
```
Comment (RyanRHall) [inline core/services/keeper/validate.go:37]: I think remove this flag. If the performUpkeep eth_call reverts, then something is wrong (it should never revert) and we want to know about it rather than fail silently.

File: core/services/pipeline/models.go [hunk 4]
```diff
@@ -50,8 +50,9 @@ type Run struct {
 	PipelineTaskRuns []TaskRun        `json:"taskRuns"`
 	State            RunStatus        `json:"state"`
 
-	Pending   bool
-	FailEarly bool
+	Pending bool
+	// FailEarlySilently implies an error has happened and a task has been marked to failEarly
+	FailEarlySilently bool
```
Comment (RyanRHall) [inline core/services/pipeline/models.go:55]: wondering if we should just change `failEarly` to `failSilently` - might be less confusing?

File: core/services/pipeline/task.conditional.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+)
+
+// ConditionalTask checks if data is false
+// for now this is all we need but in the future we can
+// expand this to handle more general conditional statements
```
Comment (RyanRHall) [inline core/services/pipeline/task.conditional.go:14]: :+1: love it

**Review by RyanRHall — CHANGES_REQUESTED** [review]

**Comment by RyanRHall** [comment]
Looks great @cmalec, just a couple naming nitpicks

File: core/services/pipeline/models.go (hunk 4 above)
Comment (cmalec) [reply core/services/pipeline/models.go:55]: in the task? idk cause it does 'fail early' as it does mark the other tasks canceled. and nothing in the task is done silently only once it hits the run level then things are done silently. 

actually I am going to change FailEarlySilently to just FailSilently because once it hits the run level its actually not doing anything early just silently.

So the task fails early and when that bubbles to the run it fails silently. I think hopefully that makes more sense.

File: core/services/pipeline/models.go [hunk 6]
```diff
@@ -50,8 +50,9 @@ type Run struct {
 	PipelineTaskRuns []TaskRun        `json:"taskRuns"`
 	State            RunStatus        `json:"state"`
 
-	Pending   bool
-	FailEarly bool
+	Pending bool
+	// FailSilently implies an error has happened and a task has been marked to failEarly
```
Comment (infiloop2) [inline core/services/pipeline/models.go:54]: Not sure i follow. failEarly has different semantics than failSilently for me

File: integration-tests/go.sum [hunk 7]
```diff
@@ -1811,12 +1811,8 @@ github.com/slack-go/slack v0.10.2 h1:KMN/h2sgUninHXvQI8PrR/PHBUuWp2NPvz2Kr66tki4
 github.com/slack-go/slack v0.10.2/go.mod h1:5FLdBRv7VW/d9EBxx/eEktOptWygbA9K2QK/KW7ds1s=
 github.com/smartcontractkit/chainlink v0.8.10-0.20200825114219-81dd2fc95bac/go.mod h1:j7qIYHGCN4QqMXdO8g8A9dmUT5vKFmkxPSbjAIfrfNU=
 github.com/smartcontractkit/chainlink v0.9.5-0.20201207211610-6c7fee37d5b7/go.mod h1:kmdLJbVZRCnBLiL6gG+U+1+0ofT3bB48DOF8tjQvcoI=
-github.com/smartcontractkit/helmenv v1.0.60 h1:c1FDBjFn+LHv8x0NbGsci3FziZF9208C9ulbi0O7kcY=
-github.com/smartcontractkit/helmenv v1.0.60/go.mod h1:VbfetB06uItvuxJKdZ6AftpbATYCnT40IaKRr75ebcs=
```
Comment (infiloop2) [inline integration-tests/go.sum:1815]: are these related changes?

File: core/services/pipeline/models.go (hunk 4 above)
Comment (infiloop2) [reply core/services/pipeline/models.go]: I'm pretty confused by fail early and fail silently semantics. Could you add more descriptive comment here about how it affects further pipeline tasks

**Review by infiloop2 — COMMENTED** [review]
Looks great @cmalec! I'm just a bit confused by failEarly, failSilently semantics, could you add a better descriptive comment there

File: core/services/pipeline/runner.go [hunk 8]
```diff
@@ -519,13 +511,11 @@ func (r *runner) Run(ctx context.Context, run *Run, l logger.Logger, saveSuccess
 	}
 
 	for {
-		if _, err = r.run(ctx, pipeline, run, NewVarsFrom(run.Inputs.Val.(map[string]interface{})), l); err != nil {
-			return false, errors.Wrapf(err, "failed to run for spec ID %v", run.PipelineSpec.ID)
-		}
+		r.run(ctx, pipeline, run, NewVarsFrom(run.Inputs.Val.(map[string]interface{})), l)
 
 		if preinsert {
 			// if run failed and it's failEarly, skip StoreRun and instead delete all trace of it
```
Comment (infiloop2) [inline core/services/pipeline/runner.go:517]: nit. comment uses failEarly

File: core/services/keeper/validate.go [hunk 9]
```diff
@@ -33,6 +33,18 @@ decode_check_upkeep_tx   [type=ethabidecode
...
+                          extractRevertReason=true
+                          evmChainID="$(jobSpec.evmChainID)"
+                          contract="$(jobSpec.contractAddress)"
+                          from="$(jobSpec.fromAddress)"
+                          gas="$(jobSpec.performUpkeepGasLimit)"
+                          data="$(encode_perform_upkeep_tx)"]
+decode_check_perform_tx  [type=ethabidecode
+                          abi="bool success"]
+check_success            [type=conditional
+                          failEarly=true
```
Comment (infiloop2) [inline core/services/keeper/validate.go:46]: nit. failSilently

File: core/services/pipeline/models.go (hunk 4 above)
Comment (cmalec) [reply core/services/pipeline/models.go]: I tried to improve the comments. failEarly is a flag on a ethCall task. if an ethcall task with the failEarly flag fails during a pipeline run then it cancels the rest of the jobs and bubbles up to the run where the failSilently flag is flipped so the run knows to cleaning things up and not store the failed run in the db.

**Review by infiloop2 — APPROVED** [review]
lgtm! 

Going forward, once we have some integration tests setup we should be able to add tests changes like these: deploy contract where performUpkeep reverts, expect it to not be called, change some conditions and expect it to be called.

**Review by begmaroman — DISMISSED** [review]

File: core/services/keeper/validate.go [hunk 10]
```diff
@@ -33,6 +33,18 @@ decode_check_upkeep_tx   [type=ethabidecode
 encode_perform_upkeep_tx [type=ethabiencode
                           abi="performUpkeep(uint256 id, bytes calldata performData)"
                           data="{\"id\": $(jobSpec.upkeepID),\"performData\":$(decode_check_upkeep_tx.performData)}"]
+simulate_perform_upkeep_tx  [type=ethcall
+                          extractRevertReason=true
+                          evmChainID="$(jobSpec.evmChainID)"
+                          contract="$(jobSpec.contractAddress)"
+                          from="$(jobSpec.fromAddress)"
+                          gas="$(jobSpec.performUpkeepGasLimit)"
+                          data="$(encode_perform_upkeep_tx)"]
```
Comment (infiloop2) [inline core/services/keeper/validate.go:42]: Just thought about this, should we also add 
```
                    gasPrice="$(jobSpec.gasPrice)"
                    gasTipCap="$(jobSpec.gasTipCap)"
                    gasFeeCap="$(jobSpec.gasFeeCap)"
```
to simulate the gas price as we do in checkUpkeep

**Comment by infiloop2** [comment]
Do we also need a migration for old observation sources in the DB?

File: core/services/keeper/validate.go (hunk 10 above)
Comment (cmalec) [reply core/services/keeper/validate.go:42]: @RyanRHall i forgot why we left this out

File: core/services/keeper/upkeep_executer_test.go [hunk 11]
```diff
@@ -279,41 +299,22 @@ func Test_UpkeepExecuter_PerformsUpkeep_Happy(t *testing.T) {
...
-			Return(txmgr.EthTx{}, nil).
-			Run(func(mock.Arguments) { etxs[1].ItHappened() })
-
-		executer.OnNewLongestChain(context.Background(), head)
-		runs = cltest.WaitForPipelineComplete(t, 0, job.ID, 2, 5, jpv2.Jrm, time.Second, 100*time.Millisecond)
-		require.Len(t, runs, 2)
-		assert.False(t, runs[1].HasErrors())
-		etxs[1].AwaitOrFail(t)
-		waitLastRunHeight(t, db, upkeep, 40)
-
```
Comment (infiloop2) [inline core/services/keeper/upkeep_executer_test.go:316]: not sure why we are not testing this

Comment (cmalec) [reply core/services/keeper/upkeep_executer_test.go:316]: this test description was about skipping blocks which is still tested in the first half. but the second half was testing the logic of the old turn taking and didn't really make sense with the new turn taking. It also relied on the grace period being 0 to force an unrealistic single keeper scenario. This lead it to became flaky because it was really only passing due to its speedy looping of blocks.

File: core/services/keeper/validate.go (hunk 10 above)
Comment (RyanRHall) [reply core/services/keeper/validate.go:42]: I would vote to leave this out of the simulation. I think adding gas prices to these eth_calls was a hacky feature-add and has been buggy and cumbersome. I'd also like to remove this "back door" feature entirely in the next release of keepers and instead pas gas price explicitly as an argument to checkUpkeep. As such, I'd prefer we don't get customers accustomed to using it.

**Review by RyanRHall — APPROVED** [review]

**Review by infiloop2 — APPROVED** [review]
